packages/exp/src/commands/upload/BaseUploader.js

```diff
@@ -0,0 +1,249 @@
+    };
+  };
+
+  latestGetter = async () => {
+    const platform = this.getPlatform();
+    const [build] = await BuildInformation.getBuildInformation(
+      { limit: 1, platform },
+      this.projectDir
+    );
+    if (!build) {
+      throw new Error(`There is no compiled builds for ${platform}`);
+    }
+    return {
+      id: build.id,
+      platform,
+      remoteUrl: build.artifacts.url,
+    };
+  };
+
+  chooseBuild = async () => {
+    const listElement = build => {
+      const platformPart = build.platform === 'ios' ? 'iOS' : 'Android';
+      const message = `### ${platformPart} | ${UrlUtils.constructBuildLogsUrl(build.id)} ###`;
+
+      return {
+        name: message,
+        value: build,
+      };
+    };
+    const platform = this.getPlatform();
+    const builds = await BuildInformation.getBuildInformation(
+      { platform, limit: 10 },
+      this.projectDir
+    );
+    if (!builds.length) {
+      throw new Error(`There is no compiled builds for ${this.getPlatformName()}`);
+    }
+    const { build } = await prompt({
+      name: 'build',
+      message: 'Choose build to upload',
+      type: 'list',
+      choices: builds.map(listElement),
+    });
+
+    return {
+      id: build.id,
+      platform: build.platform,
+      remoteUrl: build.artifacts.url,
+    };
+  };
+
+  nothingGetter = async () => {
+    const SOURCES_LIST = [
+      {
+        name: 'Latest build',
+        value: 'latest',
+      },
+      {
+        name: 'Other build',
+        value: 'other',
+      },
+    ];
+    const { source } = await prompt({
+      name: 'source',
+      message: 'Which build do you want to upload?',
+      type: 'list',
+      choices: SOURCES_LIST,
+    });
+
+    if (source === 'latest') return this.latestGetter();
+    else return this.chooseBuild();
+  };
+
+  getSource = () => {
+    const options = this.options;
+    let getter;
+    if (options.path) getter = this.pathGetter;
+    else if (options.latest) getter = this.latestGetter;
+    else if (options.id) getter = this.idGetter;
+    else getter = this.nothingGetter;
+
+    return {
+      getData: getter,
+    };
+  };
+
+  getConfigData = async () => {
+    const { exp } = await ProjectUtils.readConfigJsonAsync(this.projectDir);
+    const configName = await ProjectUtils.configFilenameAsync(this.projectDir);
+    if (!exp) {
+      throw new Error(`Couldn't read ${configName} file in project at ${this.projectDir}`);
+    }
+
+    return {
+      name: exp.name,
+      android: exp.android,
+      ios: exp.ios,
+      version: exp.version,
+      debug: { configName },
+    };
+  };
+
+  copyFile = async file => {
+    if (!fs.existsSync(file)) {
+      throw new Error(`File ${file} doesn't exist`);
+    }
+    const suffix = this.getPlatform() === 'ios' ? 'ipa' : 'apk';
+    const dest = `./build.${suffix}`;
+    await fs.copyFileSync(file, dest);
+    return dest;
+  };
+
+  downloadFileFrom = async remoteUrl => {
```

Comment:
Maybe we should remove downloaded/copied file after uploading it to the store?

Reply:
This function seems to do to much. Please break it into smaller functions with clear resposibilities.